eng/common/scripts/Helpers/PSModule-Helpers.ps1

```diff
@@ -0,0 +1,97 @@
+  if ($AzModuleCachPath -and $env.PSModulePath -notcontains $AzModuleCachPath) {
+    $modulePaths += $AzModuleCachPath
+  }
+
+  $env:PSModulePath = $modulePaths -join $moduleSeperator
+
+  # Find the path that is under user home directory
+  $homeDirectories = $modulePaths.Where({ $_.StartsWith($home) })
+  if ($homeDirectories.Count -gt 0) {
+    $global:CurrentUserModulePath = $homeDirectories[0]
+    if ($homeDirectories.Count -gt 1) {
+      Write-Verbose "Found more then one module path starting with $home so selecting the first one $global:CurrentUserModulePath"
+    }
+
+    # In some cases the directory might not exist so we need to create it otherwise caching an empty directory will fail
+    if (!(Test-Path $global:CurrentUserModulePath)) {
+      New-Item $global:CurrentUserModulePath -ItemType Directory > $null
+    }
+  }
+  else {
+    Write-Error "Did not find a module path starting with $home to set up a user module path in $env:PSModulePath"
+  }
+}
+
+# If we want to use another default repository other then PSGallery we can update the default parameters
+function Install-ModuleIfNotInstalled($moduleName, $version, $repositoryUrl = $DefaultPSRepositoryUrl)
+{
+  # Check installed modules
+  $modules = (Get-Module -ListAvailable $moduleName)
+  if ($version -as [Version]) {
+    $modules = $modules.Where({ [Version]$_.Version -ge [Version]$version })
+  }
+
+  if ($modules.Count -eq 0)
+  {
+    $repositories = (Get-PSRepository).Where({ $_.SourceLocation -eq $repositoryUrl })
+    if ($repositories.Count -eq 0)
+    {
+      Register-PSRepository -Name $repositoryUrl -SourceLocation $repositoryUrl -InstallationPolicy Trusted
+      $repositories = (Get-PSRepository).Where({ $_.SourceLocation -eq $repositoryUrl })
+      if ($repositories.Count -eq 0) {
+        Write-Error "Failed to registory package repository $repositoryUrl."
+        return
+      }
+    }
+    $repository = $repositories[0]
+
+    if ($repository.InstallationPolicy -ne "Trusted") {
+      Set-PSRepository -Name $repository.Name -InstallationPolicy "Trusted"
+    }
+
+    Write-Host "Installing module $moduleName with min version $version from $repositoryUrl"
+    # Install under CurrentUser scope so that the end up under $CurrentUserModulePath for caching
+    Install-Module $moduleName -MinimumVersion $version -Repository $repository.Name -Scope CurrentUser -Force
+
+    # Ensure module installed
+    $modules = (Get-Module -ListAvailable $moduleName)
+    if ($version -as [Version]) {
+      $modules = $modules.Where({ [Version]$_.Version -ge [Version]$version })
+    }
+
+    if ($modules.Count -eq 0) {
+      Write-Error "Failed to install module $moduleName with version $version"
+      return
+    }
+  }
+
+  Write-Host "Using module $($modules[0].Name) with version $($modules[0].Version)."
+  return $modules[0]
+}
+
+Update-PSModulePath
```

Comment:
This doesn't allow dot-sourcing the script without also executing it. I wouldn't overload the intent here. If you're going to dot-source (though modules have more utility than dot-sourcing) support that, and then call `Update-PSModulePath` after dot-sourcing. Makes it more robust and reusable.

Reply:
I made the Update-PSModulePath idempotent so it shouldn't hurt if you call it multiple times which is why I was comfortable doing this as opposed to requiring all importers to call it explicitly.